Check to ensure focus has intentionally left the wrapped component in `withFocusOutside` HOC

getdave opened the pull request:
`withFocusOutside` HOC aims to detect when a focus occurs outside of the wrapped compomnent. 

Previously there was no check to see whether the blur event occurred _within_ the wrapped component. 

If it occurs _within_ then we do not want to trigger the `handleFocusOutside` callback because (by definition) the focus has not moved outside the wrapped element.

This may have been a regression introduced by removal of `react-click-outside` in https://github.com/WordPress/gutenberg/pull/16878

## How has this been tested?
Still working on this.

## Types of changes

 Bug fix (non-breaking change which fixes an issue

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Adds check to determine whether event occured within element trapping focus

Previously there was no check to see whether the blur event occured within the wrapped component. If it occurs within then we do not want to trigger the `handleFocusOutside` callback because (by definition) the focus has not moved outside the wrapped element.

This may have been a regression introduced in https://github.com/WordPress/gutenberg/pull/16878

youknowriad commented on packages/components/src/higher-order/with-focus-outside/index.js:
```diff
@@ -79,7 +79,8 @@ export default createHigherOrderComponent(
 
 				this.blurCheckTimeout = setTimeout( () => {
-					if ( 'function' === typeof this.node.handleFocusOutside ) {
+					const eventTargetElIsWithinComponent = this.node.containerRef.current && this.node.containerRef.current.contains( event.target );
```
Can we add a comment to explain the reasoning maybe? The focus being moved to another element inside the modal which striggers the "blur" event?

getdave replied:
It's actually the focus being moved outside the modal from an element _within_ the Modal that fires blur. Elements within the modal should be captured.

youknowriad replied:
So there's something I'm missing if this blurCheckTimeout is only called when we move outside the modal, there's no need for an extra check, we should call `handleFocusOutside` shouldn't we?

getdave replied:
Sorry Riad, I'm being really unclear here. 

What I mean is there is a `blur` event firing on a `button` within the Modal which causes the block of code above to run when it shouldn't.

After I can into a few issues today I'm actually starting to think this might be a problem with our code. Let me come back to you.

getdave replied:
Ok I've experimented more and I think perhaps our issue was due to the focus going down into a deeper DOM level. 

Using [an enhanced version of `Disabled`](https://github.com/WordPress/gutenberg/pull/17064) seems to resolve this issue when I use the callback to cancel the ability to tab into anchors.

youknowriad replied:
That still feels like a bug and your fix like the correct fix to me.

If a "blur" event is triggered but the focus stays inside the Modal (another element of the modal), the onFocusOutside shouldn't be triggered.

getdave replied:
@youknowriad Ok with some more digging I've narrowed down the issue to be due to `blur` events which fire when the document loses focus. This can happen when 

* switching to devtools
* switching to another application on your computer (eg: changing focus from Firefox and back to Chrome again)

When this happens you can check whether the document has focus with the apty named method `document.hasFocus()`.

I've added a test which checks whether the document still has focus before considering the blur event to be a valid case of "focus outside". If the document isn't focused we `preventDefault` to preserve the existing focused element.

Testing in Chrome and FF and seems to solve the issue.

Also wrote tests to cover this. Not sure about my process of mocking the `document.hasFocus` method so please could you validate it is kosher? Much appreciated.

getdave replied:
Another alternative might be using the Page Visibility API to monitor whether the page is focused and if it is, then assign a flag to block the calls to `handleFocusOutside()`.

```
// pseudo code high level example
document.addEventListener('visibilitychange', setFlagToBlockFocusOutside, false);
```

youknowriad commented:
Looks like that there's a lot of failing tests.

getdave commented:
Yeh I should look at those tests now.

getdave committed:
Fix handling blur event when document is not actually focused

Previously we fixed the wrong problem. The actual issue is that previously we were handling document blur events as true “focus outside” events when in fact they should be ignored as if the document is not focused then presumably the user did not intent the blur.

getdave committed:
Update to test for document loss of focus scenario

youknowriad reviewed (commented):
This works as intended but it's not fixing all the issues I hoped it would fix :) 
It's not fixing #17042 but it's slightly related.

getdave commented:
> This works as intended but it's not fixing all the issues I hoped it would fix :)
> It's not fixing #17042 but it's slightly related.

I'll take a look at that other bug as I might have some insight now I understand this area of the code better.

Thanks for approving review 👍

aduth commented on packages/components/src/higher-order/with-focus-outside/index.js:
```diff
@@ -79,6 +79,14 @@ export default createHigherOrderComponent(
+					// this blur event thereby leaving focus in place.
+					// https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus.
+					if ( ! document.hasFocus() ) {
+						event.preventDefault();
```
Is this line necessary, or is the fix contained in just avoiding the callback?

Notably, the `focus` event is not cancellable, so it doesn't seem like this should be doing anything.

See: https://developer.mozilla.org/en-US/docs/Web/API/Element/focus_event
